From 7fdd320512e1379468ad7363363e8b8aa21e2ad9 Mon Sep 17 00:00:00 2001
From: Tianling Shen <cnsztl@immortalwrt.org>
Date: Mon, 7 Aug 2023 01:20:15 +0800
Subject: [PATCH] rockchip: add NanoPC T6 support

Hardware
--------
RockChip RK3588 ARM64 (8 cores)
4/8/16GB LPDDR4X RAM
2x 2500 Base-T (PCIe, rtl8125b)
2 LEDs (POWER / WAN)
32/64/256GB eMMC on-board
Micro-SD Slot
USB 2.0 Port
Headphone
USB 3.0
USB PD 2.0
M.2 M-Key
M.2 E-Key
DC Jack 12V

Installation
------------
Uncompress the ImmortalWrt sysupgrade and write it to a micro SD card or
internal eMMC using dd.

Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>
---
 .../armv8/base-files/etc/board.d/01_leds      |    3 +
 .../armv8/base-files/etc/board.d/02_network   |    4 +
 .../etc/hotplug.d/net/40-net-smp-affinity     |    4 +
 .../boot/dts/rockchip/rk3588-nanopc-t6.dts    | 1106 +++++++++++++++++
 target/linux/rockchip/image/armv8.mk          |    9 +
 .../210-rockchip-rk356x-add-support-for-new-boards.patch        |    3 +-
 6 files changed, 1128 insertions(+), 1 deletion(-)
 create mode 100644 target/linux/rockchip/files/arch/arm64/boot/dts/rockchip/rk3588-nanopc-t6.dts

diff --git a/target/linux/rockchip/armv8/base-files/etc/board.d/01_leds b/target/linux/rockchip/armv8/base-files/etc/board.d/01_leds
index 340beeaa8cdc3..19cac6c84897a 100755
--- a/target/linux/rockchip/armv8/base-files/etc/board.d/01_leds
+++ b/target/linux/rockchip/armv8/base-files/etc/board.d/01_leds
@@ -9,6 +9,9 @@ boardname="${board##*,}"
 board_config_update
 
 case $board in
+friendlyarm,nanopc-t6)
+	ucidef_set_led_netdev "wan" "WAN" "green:wan" "eth2"
+	;;
 friendlyarm,nanopi-r2c|\
 friendlyarm,nanopi-r2s|\
 xunlong,orangepi-r1-plus|\
diff --git a/target/linux/rockchip/armv8/base-files/etc/board.d/02_network b/target/linux/rockchip/armv8/base-files/etc/board.d/02_network
index df373c3234188..f7be02c8b9fb0 100755
--- a/target/linux/rockchip/armv8/base-files/etc/board.d/02_network
+++ b/target/linux/rockchip/armv8/base-files/etc/board.d/02_network
@@ -24,6 +24,9 @@ rockchip_setup_interfaces()
 	xunlong,orangepi-r1-plus-lts)
 		ucidef_set_interfaces_lan_wan 'eth1' 'eth0'
 		;;
+	friendlyarm,nanopc-t6)
+		ucidef_set_interfaces_lan_wan 'eth1' 'eth2'
+		;;
 	fastrhino,r66s|\
 	firefly,rk3568-roc-pc|\
 	friendlyarm,nanopi-r5c|\
@@ -83,6 +86,7 @@ rockchip_setup_macs()
 	fastrhino,r66s|\
 	fastrhino,r68s|\
 	firefly,rk3568-roc-pc|\
+	friendlyarm,nanopc-t6|\
 	friendlyarm,nanopi-r2c|\
 	friendlyarm,nanopi-r2s|\
 	hinlink,opc-h66k|\
diff --git a/target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity b/target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity
index f605ed00d2c44..58c1599fd7dc8 100644
--- a/target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity
+++ b/target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity
@@ -34,6 +34,10 @@ firefly,rk3568-roc-pc)
 	set_interface_core 2 "eth0"
 	set_interface_core 4 "eth1"
 	;;
+friendlyarm,nanopc-t6)
+	set_interface_core 2 "eth1"
+	set_interface_core 4 "eth2"
+	;;
 friendlyarm,nanopi-r2c|\
 friendlyarm,nanopi-r2s|\
 seewo,sv21-rk3568|\
diff --git a/target/linux/rockchip/image/armv8.mk b/target/linux/rockchip/image/armv8.mk
index 7f9e3c19ce4..2275130e3d1 100644
--- a/target/linux/rockchip/image/armv8.mk
+++ b/target/linux/rockchip/image/armv8.mk
@@ -49,6 +49,15 @@ define Device/friendlyarm_nanopc-t4
 endef
 TARGET_DEVICES += friendlyarm_nanopc-t4
 
+define Device/friendlyarm_nanopc-t6
+  DEVICE_VENDOR := FriendlyARM
+  DEVICE_MODEL := NanoPC T6
+  SOC := rk3588
+  BOOT_FLOW := pine64-img
+  DEVICE_PACKAGES := kmod-r8125
+endef
+TARGET_DEVICES += friendlyarm_nanopc-t6
+
 define Device/friendlyarm_nanopi-r2c
   DEVICE_VENDOR := FriendlyARM
   DEVICE_MODEL := NanoPi R2C
